chia/server/node_discovery.py

```diff
@@ -335,33 +365,16 @@ async def _connect_to_peers(self, random) -> None:
                     disconnect_after_handshake = True
                     retry_introducers = False
                 initiate_connection = self._num_needed_peers() > 0 or has_collision or is_feeler
-                client_connected = False
-                if addr is not None and initiate_connection:
-                    try:
-                        client_connected = await self.server.start_client(
-                            addr,
-                            is_feeler=disconnect_after_handshake,
-                            on_connect=self.server.on_connect,
-                        )
-                    except Exception as e:
-                        self.log.error(f"Exception in create outbound connections: {e}")
-                        self.log.error(f"Traceback: {traceback.format_exc()}")
-
-                    if self.server.is_duplicate_or_self_connection(addr):
-                        # Mark it as a softer attempt, without counting the failures.
-                        await self.address_manager.attempt(addr, False)
-                    else:
-                        if client_connected is True:
-                            await self.address_manager.mark_good(addr)
-                            await self.address_manager.connect(addr)
-                        else:
-                            await self.address_manager.attempt(addr, True)
-
                 sleep_interval = 1 + len(groups) * 0.5
                 sleep_interval = min(sleep_interval, self.peer_connect_interval)
                 # Special case: try to find our first peer much quicker.
                 if len(groups) == 0:
                     sleep_interval = 0.1
+                if addr is not None and initiate_connection:
+                    while len(self.pending_outbound_connections) >= MAX_CONCURRENT_OUTBOUND_CONNECTIONS:
```

Comment:
we also have a limit on the number outgoing peer connections, defaulting to 10, or something like that. It would seem reasonable to also limit this to the number of outgoing connection "slots" we have available. Otherwise we may end up connecting to 100 peers, and then disconnect 90 of them because we exceeded this limit.

In order to increase the connection attempt rate, we could limit the number of attempts to 1.5 times this limit, to over commit a little bit.

Reply:
the problem with limiting to 15 peers is we have a 30s timeout. if all 15 peers tried are unavailable, this will block discovering peers for 30 seconds. (just because of the timeout). after we get our 1st peer, the sleep_interval starts doing indirectly that - peer discovery sleeps for 3s at least between tries, having at most 30 / 3 = 10 peers attempted before timeouts expire. (the more peers we have the greater sleep_interval gets)

the purpose of the loop is limiting the number of connections made when we have 0 peers and we sleep 0.1s, but in all other cases it should have minimal impact.

Reply:
probably there are cases where before we handshake to our first peer, we are lucky to initiate connections with other good peers - but after all settle, sleep_interval will handle this properly.

Reply:
Do you have a sense of how common it is for a connection attempt to fail because it times out after 30 seconds vs. failing because of a "connection refused" or "host unreachable" error?

Reply:
I understand that limiting the connection attempts to the number of open slots for outgoing connections would be too low a limit. But it doesn't seem 100 is a reasonable limit either, is it?

The perfect over-commit would be based on the probability of a connection succeeding. If 100 seems to work well right now, I'm worried that it may be too high in the future. For instance, the over-commit should probably be proportional to how many slots we still have to fill. Say we have 8 slots for outgoing connections, and we currently have 7 peers that we connected to. We effectively only have 1 slot left. Do we really need to initiate 100 connection attempts to fill that slot?